tests/integ/test_reduce_describe_query.py

```diff
@@ -0,0 +1,119 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+from typing import List
+
+import pytest
+
+from snowflake.snowpark._internal.analyzer.expression import Attribute
+from snowflake.snowpark.functions import col
+from snowflake.snowpark.session import (
+    _PYTHON_SNOWPARK_REDUCE_DESCRIBE_QUERY_ENABLED,
+    Session,
+)
+from tests.integ.utils.sql_counter import SqlCounter
+from tests.utils import IS_IN_STORED_PROC
+
+pytestmark = [
+    pytest.mark.skipif(
+        "config.getoption('local_testing_mode', default=False)",
+        reason="Reducing describe queries is not supported in Local Testing",
+    ),
+]
+
+
+@pytest.fixture(scope="module", autouse=True)
+def setup(session):
+    is_reduce_describe_query_enabled = session.reduce_describe_query_enabled
+    session.reduce_describe_query_enabled = True
+    yield
+    session.reduce_describe_query_enabled = is_reduce_describe_query_enabled
+
+
+# TODO SNOW-1728988: add more test cases with select after caching attributes on SelectStatement
+# Create from SQL
+create_from_sql_funcs = [
+    lambda session: session.sql("SELECT 1 AS a, 2 AS b"),
+]
+
+# Create from Values
+create_from_values_funcs = []
+
+# Create from Table
+create_from_table_funcs = [
+    lambda session: session.create_dataframe(
+        [[1, 2], [3, 4]], schema=["a", "b"]
+    ).cache_result(),
+]
+
+# Create from SnowflakePlan
+create_from_snowflake_plan_funcs = [
+    lambda session: session.create_dataframe([[1, 2], [3, 4]], schema=["a", "b"])
+    .group_by("a")
+    .count(),
+    lambda session: session.create_dataframe([[1, 2], [3, 4]], schema=["a", "b"]).join(
+        session.sql("SELECT 1 AS a, 2 AS b")
+    ),
+    lambda session: session.create_dataframe(
+        [[1, 2], [3, 4]], schema=["a", "b"]
+    ).rename({"b": "c"}),
+]
+
+metadata_no_change_df_ops = [
+    lambda df: df.filter(col("a") > 2),
+    lambda df: df.filter((col("a") - 2) > 2),
+    lambda df: df.sort(col("a").desc()),
+    lambda df: df.sort(-col("a")),
+    lambda df: df.limit(2),
+    lambda df: df.filter(col("a") > 2).sort(col("a").desc()).limit(2),
+    lambda df: df.sample(0.5),
+    lambda df: df.sample(0.5).filter(col("a") > 2),
```

Comment:
i think you can still have some selectStatement with case where flatten is not triggered

Reply:
yes, I have some like `lambda df: df.sort(col("a").desc()).limit(2).filter(col("a") > 2)` that flatten is not triggred will be added in next PR